bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AutoRecoveryMain.java

```diff
@@ -266,45 +302,93 @@ private static ServerConfiguration parseArgs(String[] args)
-                    if (httpServer != null && httpServer.isRunning()) {
-                        httpServer.stopServer();
-                    }
-                    LOG.info("Shutdown AutoRecoveryMain successfully");
-                }
-            });
-            LOG.info("Register shutdown hook successfully");
-            autoRecoveryMain.join();
-            System.exit(autoRecoveryMain.getExitCode());
+            server = buildAutoRecoveryServer(new BookieConfiguration(conf));
         } catch (Exception e) {
-            LOG.error("Exception running AutoRecoveryMain : ", e);
-            System.exit(ExitCode.SERVER_EXCEPTION);
+            LOG.error("Failed to build AutoRecovery Server", e);
+            return ExitCode.SERVER_EXCEPTION;
+        }
+
+        // 2. start the server
+        try {
+            ComponentStarter.startComponent(server).get();
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            // the server is interrupted
+            LOG.info("AutoRecovery server is interrupted. Exiting ...");
+        } catch (ExecutionException ee) {
+            LOG.error("Error in bookie shutdown", ee.getCause());
+            return ExitCode.SERVER_EXCEPTION;
+        }
+        return ExitCode.OK;
+    }
+
+    static LifecycleComponentStack buildAutoRecoveryServer(BookieConfiguration conf) throws Exception {
+        LifecycleComponentStack.Builder serverBuilder = LifecycleComponentStack.newBuilder()
+                .withName("autorecovery-server");
+
+        // 1. build stats provider
+        StatsProviderService statsProviderService = new StatsProviderService(conf);
+        StatsLogger rootStatsLogger = statsProviderService.getStatsProvider().getStatsLogger("");
+
+        serverBuilder.addComponent(statsProviderService);
+        LOG.info("Load lifecycle component : {}", StatsProviderService.class.getName());
+
+        // 2. build AutoRecovery server
+        AutoRecoveryService autoRecoveryService = new AutoRecoveryService(conf, rootStatsLogger);
+
+        serverBuilder.addComponent(autoRecoveryService);
+        LOG.info("Load lifecycle component : {}", AutoRecoveryService.class.getName());
+
+        // 4. build http service
+        if (conf.getServerConf().isHttpServerEnabled()) {
+            BKHttpServiceProvider provider = new BKHttpServiceProvider.Builder()
+                    .setAutoRecovery(autoRecoveryService.getAutoRecoveryServer())
+                    .setServerConfiguration(conf.getServerConf())
+                    .setStatsProvider(statsProviderService.getStatsProvider()).build();
+            HttpService httpService = new HttpService(provider, conf, rootStatsLogger);
+
+            serverBuilder.addComponent(httpService);
+            LOG.info("Load lifecycle component : {}", HttpService.class.getName());
         }
+
+        // 5. build extra services
+        String[] extraComponents = conf.getServerConf().getExtraServerComponents();
```

Comment:
Is this the same list of extra components for the bookie?

Reply:
that probably needs to be dropped. since that argument is supposed to be used on bookies only.

Reply:
removed extraComponents